backend/remote-state/http/client.go

```diff
@@ -0,0 +1,294 @@
+package http
+
+import (
+	"bytes"
+	"crypto/md5"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+
+	multierror "github.com/hashicorp/go-multierror"
+
+	"github.com/hashicorp/terraform/state"
+	"github.com/hashicorp/terraform/state/remote"
+)
+
+// RemoteClient is used by "state/remote".State to read and write
+// blobs representing state.
+// Implements "state/remote".ClientLocker
+type RemoteClient struct {
+	client *http.Client
+
+	address       string
+	updateMethod  string
+	lockAddress   string
+	unlockMethod  string
+	lockMethod    string
+	unlockAddress string
+	username      string
+	password      string
+
+	lockID       string
+	jsonLockInfo []byte
+}
+
+// Get state file and return the payload.
+func (c *RemoteClient) Get() (*remote.Payload, error) {
+	// Convert address to type URL
+	addressURL, _ := url.Parse(c.address)
+
+	resp, err := c.httpRequest(http.MethodGet, addressURL, nil, "get state")
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	// Handle the common status codes
+	switch resp.StatusCode {
+	case http.StatusOK:
+		// Handled after
+	case http.StatusNoContent:
+		return nil, nil
+	case http.StatusNotFound:
+		return nil, nil
+	default:
+		return nil, fmt.Errorf("Unexpected HTTP response code %d", resp.StatusCode)
+	}
+
+	// Read in the body
+	buf := bytes.NewBuffer(nil)
+	if _, err := io.Copy(buf, resp.Body); err != nil {
+		return nil, fmt.Errorf("Failed to read remote state: %s", err)
+	}
+
+	// Create the payload
+	payload := &remote.Payload{
+		Data: buf.Bytes(),
+	}
+
+	md5 := md5.Sum(buf.Bytes())
+	// If there was no data, then return nil
+	if len(payload.Data) == 0 {
```

Comment:
you could move this check up after the buffer has been filled with the `io.Copy` operation:

```
buf := bytes.NewBuffer(nil)
if _, err := io.Copy(buf, resp.Body); err != nil {
    // handle
}
bufBytes := buf.Bytes()
if len(bufBytes) == 0 {
    return nil, nil
}
payload := &remote.Payload{Data: bufBytes}
md5 := md5.Sum(bufBytes)
// ...
```

This approach also gets the buffer bytes operation immediately after and "caches" it for later use. Then you don't have to keep calling `buf.Bytes()` which saves some time if that is a costly operation.